Remove unnecessary fixme message in DOMStringList::contains()

AlexsanderDamaceno opened the pull request:
#### c430a96674cebef5c520239dbbd8c2c4ce8d12b3
<pre>
Remove unnecessary fixme message in DOMStringList::contains()
<a href="https://bugs.webkit.org/show_bug.cgi?id=285160">https://bugs.webkit.org/show_bug.cgi?id=285160</a>

Reviewed by Ryosuke Niwa.

Remove unnecessary fixme message in DOMStringList::contains()

* Source/WebCore/dom/DOMStringList.cpp:
(WebCore::DOMStringList::contains const):

Canonical link: <a href="https://commits.webkit.org/289769@main">https://commits.webkit.org/289769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6876a520d6c188ac0f52133405d26dd139de4c2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87980 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7496 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42372 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92883 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38730 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90031 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15672 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67909 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25650 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90982 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/6030 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48278 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5806 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34005 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37837 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34886 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94754 "") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15148 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/94754 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75453 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/94754 "") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20394 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8152 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13719 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15166 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14908 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18353 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/16690 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/DOMStringList.h:
```diff
@@ -64,9 +65,12 @@ class DOMStringList : public RefCounted<DOMStringList> {
     }
 
     Vector<String> m_strings;
+    HashSet<String> m_stringsSet;
```
This is a trade-off memory vs look-up speed. This will need to be A/B tested on Speedometer & memory benchmarks since this increases memory usage.

cdumez replied:
Also, a HashSet + a Vector is basically a ListHashSet.

AlexsanderDamaceno replied:
Ok, in this case should use ListHashSet?

cdumez replied:
I have just started perf A/B testing for this version. But yes, we should likely test a ListHashSet too.

cdumez replied:
If you upload a version with a ListHashSet, I'll start A/B testing for that one too.

AlexsanderDamaceno replied:
Ok, should do in a different pr or could change this one?

cdumez replied:
Either is fine.

cdumez replied:
But also, you claim to make a performance optimization but provided no numbers in the commit log. What is the motivation for this change? Which benchmark does this improve? Linear search in a vector is actually very fast (as fast or faster than a HashSet) when there are only a few items. In general I wouldn't expect a lot of items in the DOMTokenList. My point is that this change is controversial. I don't mind triggering A/B testing to see if this helps any of the benchmarks we are tracking (without regressing any) but I'm curious WHY you're proposing this change, what the motivation is.

AlexsanderDamaceno replied:
You’re right, a linear scan is a good choice for now since it is managing only a few items. However, it might also be worth make this change. I Think this would not  hurt the previus  performance and, if this object starts being used to hold a large number of items, it would already be optimized. This bug was filled  in Bugzilla  by other person and I thought would be a valid optimization.

AlexsanderDamaceno replied:
Regarding the benchmark, there any tool in the project that  could use for see the results of this change?

AlexsanderDamaceno replied:
@cdumez I think would be no possible use ListHashSet because in function `DOMStringList::item`  we access  the item using the index passed as argument for the function this way `m_strings[index]`.

cdumez replied:
I guess you're right, `ListHashSet` maintains ordering but doesn't allow fast indexing. Anyway, Let's see what the perf results for your current PR show. If they progress one of the benchmarks we track, then we should do it. If not, I don't think it would be worth it.

cdumez replied:
According to A/B testing, this PR does not progress the Speedometer benchmark. However, it is a ~0.7% memory regression on macOS (Membuster benchmark). I do not think this is viable.

karlcow replied:
just for context. I filed the bug on bwo, just because of the comment in the code. 
And having real data to back it up is a good thing. Thanks @cdumez for the A/B testing. Maybe the comment should be modified or removed if it doesn't seem a good thing to do in the end.

AlexsanderDamaceno replied:
@cdumez  do you think we could remove this FIXME message?

cdumez replied:
Sure, just drop the FIXME and use something like:
```
// Currently, all consumers of DOMStringList store fairly small lists and thus an O(n) algorithm is OK.
```

AlexsanderDamaceno committed:
Remove unnecessary fixme message in DOMStringList::contains()
https://bugs.webkit.org/show_bug.cgi?id=285160

Reviewed by Ryosuke Niwa.

Remove unnecessary fixme message in DOMStringList::contains()

* Source/WebCore/dom/DOMStringList.cpp:
(WebCore::DOMStringList::contains const):

Canonical link: https://commits.webkit.org/289769@main
